etc/download.sh

```diff
@@ -1,12 +1,40 @@
-#/bin/sh
+#/usr/bin/env bash
 #
 # This script expects a single argument, which should be an URL pointing to a
 # file for download; it then tries to download that file, into a local file
 # with the same name as the remote file.
+# If a file of the same name exists in ~/.gap-package-cache, then this file
+# is used if it has the same hash
 set -e
 
 # check whether curl is available, and if so, delegate to it
-command -v curl >/dev/null 2>&1 && exec curl -L -O "$@"
+command -v curl >/dev/null 2>&1 &&
+{
+    name=${@##*/}
+    # Check if file already exists and has right hash
+    if [ -e ~/.gap-package-cache/${name} ] ; then
+        command -v sha256sum > /dev/null 2>&1 &&
+        {
+            # Retry multiple times, including if download is very slow (>10 minutes)
+            hash1=$(curl --retry 5 --retry-delay 5 --max-time 600 -L "$@.sha256")
+            hash2=$(sha256sum ~/.gap-package-cache/${name} | cut -d' ' -f1)
+            if [ "x$hash1" == "x$hash2" ] ; then
+                echo "Getting file from cache"
+                cp ~/.gap-package-cache/${name} .
+                exit 0
+            fi;
```

Comment:
I don't think the semicolon is needed here, is it?
```suggestion
            fi
```